File: Video Downloader.py
import yt_dlp as youtube_dl
import urllib.request
from PIL import Image
from docx import Document



class vid_downloader:

    def __init__(self, language, path, urls, vid_or_aud):
        self.language = language
        self.path = path
        self.urls = urls
        self.vid_or_aud = vid_or_aud
        self.info = [[],[]]

    def input_info(self, video_url):

        with youtube_dl.YoutubeDL() as ydl:

            video_info = ydl.extract_info(video_url, download=False)

            # Obtaining title, description, and thumbnail url
            title = video_info.get('title', None)
            description = video_info.get("description", None)

            self.info[0].append(title)
            self.info[1].append(description)
       
    def get_vid_info(self):
        return self.info

    def download_video(self, video_url, video_count, language):    
        
        # Download the video using youtube-dl
        # Program Configurations
        #   downloads in mp4
        #   downloads to specific folder
        #   provides general video information (title, description, etc.)
        ydl_opts = {
            'noplaylist': True,
            'format':'mp4',
            'outtmpl': self.path + f'Psych2go {language} Video {video_count}.%(ext)s' 
        }

        with youtube_dl.YoutubeDL(ydl_opts) as ydl:
            video_info = ydl.extract_info(video_url, download=False)
            
            # Obtaining title, description, and thumbnail url
            title = video_info.get('title', None)
            description = video_info.get("description", None)
            thumbnail_url = video_info.get('thumbnail', None)

            # Downloading Video
            print("\nDownloading vid...")
            ydl.download([video_url])
            print("Download completed!!")

            # Downloading thumbnail
            self.download_thumbnail(thumbnail_url, self.path, video_count, language)

            # Formatting Thumbnail
            self.edit_thumbnail(video_count, self.path, language)

            # Writing title and description to a document
            self.save_info(title, description, video_count, self.path)

    def download_audio(self, video_url, video_count, language):
        
        # Download the video using youtube-dl
        # Program Configurations
        #   downloads in mp3
        #   downloads to specific folder
        #   provides general video information (title, description, etc.)
        ydl_opts = {
            'noplaylist': True,
            'verbose': True,
            'format': 'bestaudio/best',
            'outtmpl': self.path + f'Psych2go {language} Audio {video_count}.%(ext)s',
            'postprocessors': [{
                'key': 'FFmpegExtractAudio',
                'preferredcodec': 'mp3',
                'preferredquality': '192',
            }]  
        }

        with youtube_dl.YoutubeDL(ydl_opts) as ydl:

            # Downloading Video
            print("\nDownloading audio...")
            ydl.download([video_url])
            print("Download completed!!")

    def download_thumbnail(self, thumbnail_url, path, num, language):

        full_path = path + f"{language} Video Thumbnail {num}" + ".jpg"

        print("\nDownloading Thumbnail...")
        urllib.request.urlretrieve(thumbnail_url, full_path)
        print("Download completed!!")

    def edit_thumbnail(self, num, path, language):

        print("\nEditing Thumbnail...")
        image = Image.open(path + f"{language} Video Thumbnail {num}.jpg")
        image = image.resize((1000, 1050))
        image.save(path + f"{language} Video Thumbnail {num}.jpg")
        print("Thumbnail Edited!!")

    def save_info(self, title, description, num, path):
        
        doc = Document(path + 'Psych2goInfo.docx')
        
        print("\nUploading Info...")
        doc.add_paragraph(f"Title {num}:\n\n " + title + "\n\n")
        doc.add_paragraph(f"Description {num}:\n\n " + description + "\n\n")
        doc.add_page_break()
        print("Upload Finished!!")
        doc.save(path + 'Psych2goInfo.docx')

    def video_or_audio(self, vid_or_aud):
        if vid_or_aud == 1:
            for url in self.urls:
                video_count = self.urls.index(url) + 1
                self.download_video(url, video_count, self.language)
                self.input_info(url)

        elif vid_or_aud == 2:
            for url in self.urls:
                video_count = self.urls.index(url) + 1
                self.download_video(url, video_count, self.language)
                self.download_audio(url, video_count, self.language)
                self.input_info(url)

File: videodownloader.js
const ytdl = require('youtube-dl-exec');
const axios = require('axios');
const fs = require('fs');
const sharp = require('sharp');
const fsp = require('fs/promises');
const Canvas = require('canvas');
const jimp = require('jimp');
const { resolve } = require('path');

const delay = (ms) => new Promise(resolve => setTimeout(resolve, ms));

class VideoDownloader {
    
    constructor(language, path) {
        this.language = language;
        this.path = path;
    }

    async getVidInfo(url) {
        try {
            const options = {
                dumpSingleJson: true,
                skipDownload: true
            };

            console.log("Obtaining Video Info...");

            return await ytdl(url, options);
        } catch (error) {console.error('Error while obtaining video information: ', error);};
    }

    async downloadVideo(url, count) {
        const options = {
            format: 'bestvideo[ext=mp4]+bestaudio[ext=m4a]/best[ext=mp4]/best',
            output: this.path + `Psych2go ${this.language} Video ${count}.mp4`
        };

        try {
            console.log("Downloading Video...");
            await ytdl(url, options);
            console.log("Video Downloaded!");
        } catch (error) {console.error("Error while downloading Video: ", error);}
    }

    async downloadAudio(url, count) {
        const options = {
            format: 'bestaudio[ext=mp3]/bestaudio/best',
            output: this.path + `Psych2go ${this.language} Audio ${count}.mp3`
        };

        try {
            console.log("Downloading Audio...");
            await ytdl(url, options);
            console.log("Audio Downloaded!");
        } catch (error) {console.error("Error while downloading Audio: ", error);}
    }

    async downloadThumbnail(url, count) {
        try {
            const imagePath = this.path + `${this.language} Video Thumbnail ${count}.jpeg`;

            console.log('Downloading Thumbnail...');

            const info = await this.getVidInfo(url);

            const response = await axios({method: 'GET', url: info.thumbnail, responseType: 'stream'});

            response.data.pipe(fs.createWriteStream(imagePath));

            return new Promise((resolve, reject) => {
                response.data.on('end', () => {resolve()}) 
                response.data.on('error', () => {reject(error)})           
            });

        } catch (error) {console.error('Error while downloading thumbnail: ', error);}
    }

    async editThumbnail(count) {
        const imagePath = this.path + `${this.language} Video Thumbnail ${count}.jpeg`;
        const editedImagePath = this.path + `${this.language} Video Thumbnail ${count} Edited.jpeg`;
        const width = 1000;
        const height = 1000;

        try {
            await sharp(imagePath)
                .resize(width, height, {fit: 'fill'})
                .toFile(editedImagePath);
    
            console.log(`Image resized and saved to ${editedImagePath}`);
        } catch (error) {
            console.error('Error resizing image:', error);
        }
    }

    async deleteVideo(count) {
        const filePath = this.path + `Psych2go ${this.language} Video ${count}.mp4`;
        console.log("Deleting Video...");
    
        try {
            await fsp.unlink(filePath);
            console.log("Video Deleted!");
        } catch (err) {
            console.error('Error deleting video:', err);
        }
    }

    async deleteAudio(count) {
        const filePath = this.path + `Psych2go ${this.language} Audio ${count}.mp4`;
        console.log("Deleting Audio...");
    
        try {
            await fsp.unlink(filePath);
            console.log("Audio Deleted!");
        } catch (err) {
            console.error('Error deleting audio:', err);
        }
    }

    async deleteThumbnail(count) {
        const filePath = this.path + `${this.language} Video Thumbnail ${count}.jpeg`;
        console.log("Deleting Thumbnail...");
    
        try {
            await fsp.unlink(filePath);
            console.log("Thumbnail Deleted!");
        } catch (err) {
            console.error('Final attempt failed. Error deleting Thumbnail:', err);
        }
    }

    async deleteEditedThumbnail(count) {
        const editedImagePath = this.path + `${this.language} Video Thumbnail ${count} Edited.jpeg`;
        console.log("Deleting Edited Thumbnail...");
        fs
        try {
            await fsp.unlink(editedImagePath);
            console.log("Edited Thumbnail Deleted!");
        } catch (err) {
            console.error('Error deleting Edited Thumbnail:', err);
        }
    }
}

module.exports = VideoDownloader;